hyperjumptech/monika
# Refactor the probes

# Monika Pull Request (PR)  

## Before PR
1. Probes are mixed between drivers and loggers
2. Difficult to add new probes

## After PR
1. Should be easier to add new probes
2. Code cleaned up
3. Better separation of responsibilities 

## How did you implement / how did you fix it  
1. Refactored the interface between driver and dispatcher
2. Detach drivers from logger and decoding of errors/status codes

## How to test  
1. Run `npm run test` should all pass
2. Use the following test config yaml for local test
```yaml
probes:
  # http checks
  - id: http-probe-01
    name: 'test status 400'
    interval: 5

    requests:
      - url: http://httpbin.org/status/400
        method: GET
        timeout: 5000
        saveBody: false
        headers:
          Content-Type: application/json; charset=utf-8
    alerts:
      - query: response.status != 400
        message: Http Response status code is not 200!
      - query: response.time > 10000
        message: Yo too slow!
    incidentThreshold: 2
    recoveryThreshold: 2

  - id: http-probe-02
    name: 'test good github.com'
    interval: 10

    requests:
      - url: https://github.com
    alerts:
      - query: response.status != 200
        message: Http Response status code is not 200!
      - query: response.time > 10000
        message: Yo too slow!
    incidentThreshold: 2
    recoveryThreshold: 2

# tcp check
  - id: tcp-check-01
    name: bad-tcp-ip
    alerts:
      - query: response.status == 500
        message: response status message
    incidentThreshold: 1
    recoveryThreshold: 1
    socket:
      host: 1.2.3.4
      port: 3333
      data: Hello from Monika

  - id: tcp-check-02
    name: checking ssh tcp
    alerts:
      - query: response.status == 500
        message: response status message
    incidentThreshold: 1
    recoveryThreshold: 1

    socket:
      host: 1.2.3.4 # please user your valid ssh ip
      port: 22
      data: hello

# icmp check
  - id: ping-01
    name: bad-ping-check
    description: check-ping-result
    interval: 10
    incidentThreshold: 3
    recoveryThreshold: 3
    requests:
      - method: GET
        url: cloudflare.net
        timeout: 5000
        saveBody: false
        ping: true

    alerts:
      - query: response.status != 200
        message: no ping

  - id: ping-02
    name: good-ping-check
    description: check-ping-nonexistent
    interval: 10
    incidentThreshold: 3
    recoveryThreshold: 3
    requests:
      - method: GET
        url: hyperjump.co.net
        timeout: 5000
        saveBody: false
        ping: true

    alerts:
      - query: response.status != 200
        message: no ping

notifications:
  - id: unique-id-desktop
    type: desktop
```
3. Run monika with the above config: `bin/bin-run -c monika.yaml -s -1`

## Screenshot
